models/repo_list.go

```diff
@@ -153,6 +154,11 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
 		}
 
 		cond = cond.Or(builder.And(builder.Like{"lower_name", opts.Keyword}, builder.In("owner_id", ownerIds)))
+
+		if opts.Collaborate {
+			cond = cond.Or(builder.Expr(`id IN (SELECT repo_id FROM "access" WHERE access.user_id = ? AND owner_id != ?)`, opts.Searcher.ID, opts.Searcher.ID),
```

Comment:
This is still not correct. This will match repos that do not match the `opts.Starred` or `opts.OwnerID` options.

Apologies if I wasn't clear in my previous comment. Instead of relisting all the other conditions (which is fragile if we ever add more conditions), I think we should do something like

```
// compute ownerIds, existing code up to line 154 is fine
searcherReposCond := builder.In("owner_id", ownerIds)
if opts.Collaborate {
	searcherReposCond.Or(builder.Expr(`id IN (SELECT repo_id FROM "access" WHERE access.user_id = ? AND owner_id != ?)`, opts.Searcher.ID, opts.Searcher.ID)
}
cond.And(searcherReposCond)
```